File: Complete Reference/DynamicProgramming/8-24-2023_cses_coin_comb_2.py
def recurse(coins, startIndex, remPrice):
    if startIndex >= len(coins):
        return int(remPrice == 0)
    if (remPrice - coins[startIndex]) < 0:
        return 0
    combNotTakeCurrCoin = recurse(coins, startIndex + 1, remPrice)
    combTakeCurrCoin = ((remPrice - coins[startIndex]) == 0) + recurse(
        coins, startIndex, remPrice - coins[startIndex]
    )
    totalComb = combNotTakeCurrCoin + combTakeCurrCoin
    totalComb = totalComb % 1000000007
    return totalComb


if __name__ == "__main__":
    [n, x] = list(map(int, input().split(" ")))
    coins = list(map(int, input().split(" ")))
    coins = sorted(coins)
    # print(coins, recurse(coins, 0, x))
    dp = []
    for price in range(0, x + 1):
        dp.append([])
        value = int(price == 0)
        for coin in range(0, n + 1):
            dp[price].append(value)

    for price in range(1, x + 1):
        for coin in range(1, len(coins) + 1):
            # as I am not taking the present coin I will be maintain the previous coins combinations at the present coin
            notTake = dp[price][coin - 1]
            dp[price][coin] = notTake
            if (price - coins[coin - 1]) < 0:
                break
            take = dp[price - coins[coin - 1]][coin]
            dp[price][coin] += take
            dp[price][coin] = dp[price][coin] % 1000000007
    print(dp[x][n])

File: Complete Reference/DynamicProgramming/8-24-2023_cses_coin_comb_2.cpp
#include <bits/stdc++.h>
using namespace std;
int main()
{
    int n, x;
    cin >> n >> x;
    vector<int> coins(n);
    for (int i = 0; i < n; i++)
    {
        cin >> coins[i];
    }
    sort(coins.begin(), coins.end());
    vector<vector<int>> dp(x + 1, vector<int>(n + 1));
    for (int i = 0; i < n + 1; i++)
    {
        dp[0][i] = 1;
    }
    for (int price = 1; price < (x + 1); price++)
    {
        for (int coinId = 1; coinId < (n + 1); coinId++)
        {
            int notTake = dp[price][coinId - 1];
            dp[price][coinId] = notTake;
            if ((price - coins[coinId - 1]) < 0)
                break;
            int take = dp[price - coins[coinId - 1]][coinId];
            dp[price][coinId] += take;
            dp[price][coinId] = dp[price][coinId] % 1000000007;
        }
    }
    cout << dp[x][n] << endl;
    return 0;
}
